File: src/xmss-alt/wots.c
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
// Based on the public domain XMSS reference implementation
// by Andreas Hülsing and Joost Rijneveld
/*
wots.c version 20160722
Andreas Hülsing
Joost Rijneveld
Public domain.
*/

#include <cmath>
#include <cstdint>
#include "xmss_common.h"
#include "hash.h"
#include "hash_address.h"


void wots_set_params(wots_params *params, int n, int w) {
    params->n = n;
    params->w = w;
    params->log_w = (int) log2(w);
    params->len_1 = (int) ceil(((8 * n) / params->log_w));
    params->len_2 = (int) floor(log2(params->len_1 * (w - 1)) / params->log_w) + 1;
    params->len = params->len_1 + params->len_2;
    params->keysize = params->len * params->n;
}

/**
 * Helper method for pseudorandom key generation
 * Expands an n-byte array into a len*n byte array
 * this is done using PRF
 */
void expand_seed(eHashFunction hash_func,
                 unsigned char *outseeds,
                 const unsigned char *inseed,
                 const uint32_t n,
                 const uint32_t len) {
    unsigned char ctr[32];
    for (uint32_t i = 0; i < len; i++) {
        to_byte(ctr, i, 32);
        prf(hash_func, (outseeds + (i * n)), ctr, inseed, n);
    }
}

/**
 * Computes the chaining function.
 * out and in have to be n-byte arrays
 *
 * interpretes in as start-th value of the chain
 * addr has to contain the address of the chain
 */
void gen_chain(eHashFunction hash_func,
               unsigned char *out,
               const unsigned char *in,
               unsigned int start,
               unsigned int steps,
               const wots_params *params,
               const unsigned char *pub_seed,
               uint32_t addr[8]) {
    uint32_t i, j;
    for (j = 0; j < params->n; j++)
        out[j] = in[j];

    for (i = start; i < (start + steps) && i < params->w; i++) {
        setHashADRS(addr, i);
        hash_f(hash_func, out, out, pub_seed, addr, params->n);
    }
}

/**
 * base_w algorithm as described in draft.
 *
 *
 */
void base_w(int *output,
            const int out_len,
            const unsigned char *input,
            const wots_params *params) {
    int in = 0;
    int out = 0;
    uint32_t total = 0;
    int bits = 0;
    int consumed = 0;

    for (consumed = 0; consumed < out_len; consumed++) {
        if (bits == 0) {
            total = input[in];
            in++;
            bits += 8;
        }
        bits -= params->log_w;
        output[out] = (total >> bits) & (params->w - 1);
        out++;
    }
}

void wots_pkgen(eHashFunction hash_func,
                unsigned char *pk,
                const unsigned char *sk,
                const wots_params *params,
                const unsigned char *pub_seed,
                uint32_t addr[8]) {
    uint32_t i;
    expand_seed(hash_func, pk, sk, params->n, params->len);
    for (i = 0; i < params->len; i++) {
        setChainADRS(addr, i);
        gen_chain(hash_func,
                  pk + i * params->n,
                  pk + i * params->n,
                  0,
                  params->w - 1,
                  params,
                  pub_seed,
                  addr);
    }
}


void wots_sign(eHashFunction hash_func,
               unsigned char *sig,
               const unsigned char *msg,
               const unsigned char *sk,
               const wots_params *params,
               const unsigned char *pub_seed,
               uint32_t addr[8]) {
    int basew[params->len];
    int csum = 0;
    uint32_t i = 0;

    base_w(basew, params->len_1, msg, params);

    for (i = 0; i < params->len_1; i++) {
        csum += params->w - 1 - basew[i];
    }

    csum = csum << (8 - ((params->len_2 * params->log_w) % 8));

    uint32_t len_2_bytes = ((params->len_2 * params->log_w) + 7) / 8;

    unsigned char csum_bytes[len_2_bytes];
    to_byte(csum_bytes, csum, len_2_bytes);

    int csum_basew[params->len_2];

    base_w(csum_basew, params->len_2, csum_bytes, params);

    for (i = 0; i < params->len_2; i++) {
        basew[params->len_1 + i] = csum_basew[i];
    }

    expand_seed(hash_func, sig, sk, params->n, params->len);

    for (i = 0; i < params->len; i++) {
        setChainADRS(addr, i);
        gen_chain(hash_func, sig + i * params->n, sig + i * params->n, 0, basew[i], params, pub_seed, addr);
    }
}

void wots_pkFromSig(eHashFunction hash_func,
                    unsigned char *pk,
                    const unsigned char *sig,
                    const unsigned char *msg,
                    const wots_params *wotsParams,
                    const unsigned char *pub_seed,
                    uint32_t addr[8]) {
    uint32_t XMSS_WOTS_LEN = wotsParams->len;
    uint32_t XMSS_WOTS_LEN1 = wotsParams->len_1;
    uint32_t XMSS_WOTS_LEN2 = wotsParams->len_2;
    uint32_t XMSS_WOTS_LOG_W = wotsParams->log_w;
    uint32_t XMSS_WOTS_W = wotsParams->w;
    uint32_t XMSS_N = wotsParams->n;

    int basew[XMSS_WOTS_LEN];
    int csum = 0;
    unsigned char csum_bytes[((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) + 7) / 8];
    int csum_basew[XMSS_WOTS_LEN2];
    uint32_t i = 0;

    base_w(basew, XMSS_WOTS_LEN1, msg, wotsParams);

    for (i = 0; i < XMSS_WOTS_LEN1; i++) {
        csum += XMSS_WOTS_W - 1 - basew[i];
    }

    csum = csum << (8 - ((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) % 8));

    to_byte(csum_bytes, csum, ((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) + 7) / 8);
    base_w(csum_basew, XMSS_WOTS_LEN2, csum_bytes, wotsParams);

    for (i = 0; i < XMSS_WOTS_LEN2; i++) {
        basew[XMSS_WOTS_LEN1 + i] = csum_basew[i];
    }
    for (i = 0; i < XMSS_WOTS_LEN; i++) {
        setChainADRS(addr, i);
        gen_chain(hash_func,
                  pk + i * XMSS_N, sig + i * XMSS_N,
                  basew[i], XMSS_WOTS_W - 1 - basew[i], wotsParams, pub_seed, addr);
    }
}

File: src/rust_wrapper/xmss_alt/wots.rs
use super::hash::{hash_f, prf};
use super::hash_address::{set_chain_adrs, set_hash_adrs};
use super::hash_functions::HashFunction;
use super::xmss_common::to_byte;
use std::cmp::min;

macro_rules! log2 {
    ($val:expr, $type:ty) => {
        ($val as f32).log2() as $type
    };
}

/**
 * WOTS parameter set
 *
 * Meaning as defined in draft-irtf-cfrg-xmss-hash-based-signatures-02
 */
// FIXME: Get rid of this
pub struct WOTSParams {
    pub len_1: u32,
    pub len_2: u32,
    pub len: u32,
    pub n: u32,
    pub w: u32,
    pub log_w: u32,
    pub keysize: u32,
}

impl WOTSParams {
    /**
     * Set the WOTS parameters,
     * only n and w are required as inputs,
     * len, len_1, and len_2 are computed from those.
     *
     * Assumes w is a power of 2
     */
    pub fn new(n: u32, w: u32) -> Self {
        let log_w = log2!(w, u32);
        let len_1 = ((8 * n) as f32 / (log_w as f32)).ceil() as u32;
        let len_2 = (log2!(len_1 * (w - 1), f32) / log_w as f32).floor() as u32 + 1;
        let len = len_1 + len_2;
        let keysize = len * n;
        WOTSParams {
            n,
            w,
            log_w,
            len_1,
            len_2,
            len,
            keysize,
        }
    }
}

/**
 * Helper method for pseudorandom key generation
 * Expands an n-byte array into a len*n byte array
 * this is done using PRF
 */
fn expand_seed(hash_func: &HashFunction, outseeds: &mut [u8], inseed: &[u8], n: u32, len: u32) {
    let mut ctr = vec![0 as u8; 32];
    let outseeds_length = outseeds.len();
    for i in 0..len {
        to_byte(ctr.as_mut_slice(), i.into(), 32);
        prf(
            &hash_func,
            &mut outseeds.get_mut((i * n) as usize..outseeds_length).unwrap(),
            ctr.as_mut_slice(),
            inseed,
            n,
        );
    }
}

/**
 * Computes the chaining function.
 * out and in have to be n-byte arrays
 *
 * interpretes in as start-th value of the chain
 * addr has to contain the address of the chain
 */
fn gen_chain(
    hash_func: &HashFunction,
    out: &mut [u8],
    input: &[u8],
    start: u32,
    steps: u32,
    params: &WOTSParams,
    pub_seed: &[u8],
    addr: &mut [u32; 8],
) {
    for j in 0..params.n as usize {
        out[j] = input[j];
    }

    for i in start..(min(start + steps, params.w)) {
        set_hash_adrs(addr, i);
        let mut input = vec![0; out.len()];
        input.copy_from_slice(out);
        hash_f(hash_func, out, &input, pub_seed, addr, params.n);
    }
}

/**
 * base_w algorithm as described in draft.
 */
fn base_w(output: &mut [i32], out_len: usize, input: &[u8], params: &WOTSParams) {
    let mut in_ = 0;
    let mut out = 0;
    let mut total = 0;
    let mut bits = 0;

    for _consumed in 0..out_len {
        if bits == 0 {
            total = input[in_];
            in_ += 1;
            bits += 8;
        }
        bits -= params.log_w;
        output[out] = ((total >> bits) as u32 & (params.w - 1)) as i32;
        out += 1;
    }
}

/**
 * WOTS key generation. Takes a 32byte seed for the secret key, expands it to a full WOTS secret key and computes the corresponding public key.
 * For this it takes the seed pub_seed which is used to generate bitmasks and hash keys and the address of this WOTS key pair addr
 *
 * params, must have been initialized before using wots_set params for params ! This is not done in this function
 *
 * Places the computed public key at address pk.
 */
pub fn wots_pkgen(
    hash_func: &HashFunction,
    pk: &mut [u8],
    sk: &[u8],
    params: &WOTSParams,
    pub_seed: &[u8],
    addr: &mut [u32; 8],
) {
    expand_seed(hash_func, pk, sk, params.n, params.len);
    let pk_len = pk.len();
    let mut pk_input = vec![0; pk_len];
    pk_input.copy_from_slice(pk);
    for i in 0..params.len {
        set_chain_adrs(addr, i);
        gen_chain(
            hash_func,
            pk.get_mut((i * params.n) as usize..pk_len as usize)
                .unwrap(),
            pk_input
                .get((i * params.n) as usize..pk_len as usize)
                .unwrap(),
            0,
            params.w - 1,
            params,
            pub_seed,
            addr,
        );
    }
}

pub fn wots_sign(
    hash_func: &HashFunction,
    sig: &mut [u8],
    msg: &[u8],
    sk: &[u8],
    params: &WOTSParams,
    pub_seed: &[u8],
    addr: &mut [u32; 8],
) {
    let mut basew: Vec<i32> = vec![0; params.len as usize];
    let mut csum = 0;

    base_w(&mut basew, params.len_1.try_into().unwrap(), msg, params);

    for i in 0..params.len_1 as usize {
        csum += params.w as i32 - 1 - basew[i];
    }

    csum = csum << (8 - ((params.len_2 * params.log_w) % 8));

    let len_2_bytes = ((params.len_2 * params.log_w) + 7) / 8;

    let mut csum_bytes: Vec<u8> = vec![0; len_2_bytes as usize];
    to_byte(&mut csum_bytes, csum.try_into().unwrap(), len_2_bytes);

    let mut csum_basew: Vec<i32> = vec![0; params.len_2 as usize];

    base_w(&mut csum_basew, params.len_2 as usize, &csum_bytes, params);

    for i in 0..params.len_2 {
        basew[(params.len_1 + i) as usize] = csum_basew[i as usize];
    }

    expand_seed(hash_func, sig, sk, params.n, params.len);

    for i in 0..params.len {
        set_chain_adrs(addr, i);
        let sig_length = sig.len();
        let sig_output_segment = sig.get_mut((i * params.n) as usize..sig_length).unwrap();
        let mut sig_input_segment: Vec<u8> = vec![0; sig_output_segment.len()];
        sig_input_segment.copy_from_slice(sig_output_segment);
        gen_chain(
            hash_func,
            sig_output_segment,
            &sig_input_segment,
            0,
            basew[i as usize].try_into().unwrap(),
            params,
            pub_seed,
            addr,
        );
    }
}

pub fn wots_pk_from_sig(
    hash_func: &HashFunction,
    pk: &mut [u8],
    sig: &[u8],
    msg: &[u8],
    wots_params: &WOTSParams,
    pub_seed: &[u8],
    addr: &mut [u32; 8],
) {
    let XMSS_WOTS_LEN = wots_params.len as usize;
    let XMSS_WOTS_LEN1 = wots_params.len_1 as usize;
    let XMSS_WOTS_LEN2 = wots_params.len_2 as usize;
    let XMSS_WOTS_LOG_W = wots_params.log_w as usize;
    let XMSS_WOTS_W = wots_params.w;
    let XMSS_N = wots_params.n;

    let mut basew: Vec<i32> = vec![0; XMSS_WOTS_LEN];
    let mut csum = 0;
    let mut csum_bytes: Vec<u8> = vec![0; ((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) + 7) / 8];
    let mut csum_basew: Vec<i32> = vec![0; XMSS_WOTS_LEN2];

    base_w(&mut basew, XMSS_WOTS_LEN1, msg, wots_params);

    for i in 0..XMSS_WOTS_LEN1 {
        csum += XMSS_WOTS_W - 1 - basew[i] as u32;
    }

    csum = csum << (8 - ((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) % 8));

    to_byte(
        &mut csum_bytes,
        csum.into(),
        ((XMSS_WOTS_LEN2 * XMSS_WOTS_LOG_W) as u32 + 7) / 8,
    );
    base_w(&mut csum_basew, XMSS_WOTS_LEN2, &csum_bytes, wots_params);

    for i in 0..XMSS_WOTS_LEN2 {
        basew[XMSS_WOTS_LEN1 + i] = csum_basew[i];
    }
    for i in 0..XMSS_WOTS_LEN {
        set_chain_adrs(addr, i.try_into().unwrap());
        let out = pk.get_mut(i * XMSS_N as usize..pk.len()).unwrap();
        let input = sig.get(i * XMSS_N as usize..sig.len()).unwrap();
        gen_chain(
            hash_func,
            out,
            input,
            basew[i].try_into().unwrap(),
            XMSS_WOTS_W - 1 - basew[i] as u32,
            wots_params,
            pub_seed,
            addr,
        );
    }
}
